bofire/strategies/doe/objective.py

```diff
@@ -0,0 +1,459 @@
+        X = self._convert_input_to_model_tensor(x, requires_grad=True)
+
+        # first part of jacobian
+        torch.min(
+            torch.linalg.eigvalsh(X.T @ X + self.delta * torch.eye(self.n_model_terms))
+        ).backward()
+        J1 = -1 * X.grad.detach().numpy()  # type: ignore
+        J1 = np.repeat(J1, self.n_vars, axis=0).reshape(
+            self.n_experiments, self.n_vars, self.n_model_terms
+        )
+
+        # second part of jacobian
+        J2 = self._model_jacobian_t(x)
+
+        # combine both parts
+        J = J1 * J2
+        J = np.sum(J, axis=-1)
+
+        return J.flatten()
+
+
+class KOptimality(Objective):
+    """A class implementing the evaluation of the condition number of (X.T @ X + delta)
+    and its jacobian w.r.t. the inputs. The jacobian evaluation is done analogously to
+    DOptimality with the first part of the jacobian being the jacobian of condition number
+    of (X.T @ X + delta) instead of logdet(X.T@X + delta).
+    """
+
+    def evaluate(self, x: np.ndarray) -> float:
+        """Computes condition number of (X.T@X + delta).
+        Where X is the model matrix corresponding to x.
+
+        Args:
+            x (np.ndarray): values of design variables a 1d array.
+
+        Returns:
+            cond(X.T @ X + delta)
+        """
+        X = self._convert_input_to_model_tensor(x, requires_grad=False)
+        return float(
+            torch.linalg.cond(
+                X.detach().T @ X.detach() + self.delta * torch.eye(self.n_model_terms)
+            )
+        )
+
+    def evaluate_jacobian(self, x: np.ndarray) -> np.ndarray:
+        """Computes the jacobian of the condition number of (X.T @ X + delta).
+        Where X is the model matrix corresponding to x.
+
+        Args:
+            x (np.ndarray): values of design variables a 1d array.
+
+        Returns:
+            The jacobian of cond(X.T @ X + delta) as numpy array
+        """
+        # get model matrix X
+        X = self._convert_input_to_model_tensor(x, requires_grad=True)
+
+        # first part of jacobian
+        torch.linalg.cond(
+            X.T @ X + self.delta * torch.eye(self.n_model_terms)
+        ).backward()
+        J1 = X.grad.detach().numpy()  # type: ignore
+        J1 = np.repeat(J1, self.n_vars, axis=0).reshape(
+            self.n_experiments, self.n_vars, self.n_model_terms
+        )
+
+        # second part of jacobian
+        J2 = self._model_jacobian_t(x)
+
+        # combine both parts
+        J = J1 * J2
+        J = np.sum(J, axis=-1)
+
+        return J.flatten()
+
+
+# TODO: test
+class SpaceFilling(Objective):
+
+    # TODO: how many values should be taken into account. 1 is not enough. Right now it is n_experiments, this seems to work.
+    def evaluate(self, x: np.ndarray) -> float:
+        X = self._convert_input_to_tensor(x, requires_grad=False)
+        return float(
+            -torch.sum(torch.sort(torch.pdist(X.detach()))[0][: self.n_experiments])
```

Comment:
Yes! If you use min (i.e. maximize the minimal distance) the optimization does not work properly. Even after 1000+ iterations almost no progress is done. My idea is that we don't "do enough" this way, since the gradient only gives information about the way one single pair of experiments should be moved (the one with the minimal distance).
Instead, we take the sum of the smallest n_experiments experiment distances as objective. This gives us information about how to move many more experiments while still only looking at some of the smallest distances (in total we have `n_experiments!` distances), so the optimum where we end up should not be disturbed to much by taking into account some more points. 
However, I only did tests until 100 experiments. So I am not sure yet if the scaling #considered_distances = O(n_experiments) is also good for much larger numbers of experiments.

Reply:
Oh that's brilliant! 
Yes, using min will within one step only move the closest pair apart. With your formulation all points in the `n` closest pairs are moved. 
The number of distances is `n (n - 1) / 2` (not `n!`) so the scaling is fine.

Reply:
And using the `n` closest pairs is a good choice. It's not too large, as only at the equidistant optimum, all points can simumltaneously contribute.

Reply:
Yes, you are right. `n(n-1)/2` is of course correct. Good to hear that you agree to the objective as it is now. Then I leave things as they are rn.